arbor/util/spatial_tree.hpp

```diff
@@ -0,0 +1,181 @@
+#pragma once
+
+#include "util/visit_variant.hpp"
+
+#include <arbor/common_types.hpp>
+#include <arbor/math.hpp>
+
+#include <cstddef>
+#include <limits>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace arb {
+
+// An immutable spatial data structure for storing and iterating over data in "DIM" dimensional
+// space. If DIM = 1 it's a binary tree, if DIM = 2 it's a quad tree and so on.
+template <typename T, std::size_t DIM>
+class spatial_tree {
+public:
+    static_assert(DIM >= 1, "Dimension of tree must be at least 1.");
+
+    using value_type = T;
+    using point_type = std::array<double, DIM>;
+    using node_data = std::vector<spatial_tree>;
+    using leaf_data = std::vector<T>;
+    using location_func_type = point_type (*)(const T &);
+
+    spatial_tree(): size_(0), data_(leaf_data()) {}
+
+    // Create a tree of given maximum depth and target leaf size. If any leaf holds more than the
+    // target size, it is recursively split into up to 2^DIM nodes until reaching the maximum depth.
+    // The "location" function type must have signature (const T&) -> point_type.
+    spatial_tree(std::size_t max_depth,
+        std::size_t leaf_size_target,
+        leaf_data data,
+        location_func_type location):
+        size_(data.size()),
+        data_(std::move(data)),
+        location_(location) {
+        auto &leaf_d = std::get<leaf_data>(data_);
+        if (leaf_d.empty()) return;
+
+        min_.fill(std::numeric_limits<double>::max());
+        max_.fill(-std::numeric_limits<double>::max());
```

Comment:
```suggestion
        max_.fill(std::numeric_limits<double>::min());
```

Reply:
I never knew. That would have tripped me up bad. Yet `-max` might not be representable.
>For floating-point types with denormalization, min() returns the minimum positive normalized value. Note that this behavior may be unexpected, especially when compared to the behavior of min() for integral types. To find the value that has no values less than it, use [lowest()](https://en.cppreference.com/w/cpp/types/numeric_limits/lowest).(since C++11)

`lowest` seems to be the right pick here.

Reply:
Yes, `lowest` seems to be the best fit. I've changed it now.